### PR TITLE
Remove Python 3.5 support

### LINKED ISSUE #80524: Drop support for Python 3.5
### Summary

Drop support Python 3.5.

### Issue Type

Feature Idea

### Component Name

`ansible-test`

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Require Python 2.7 or Python 3.6+ for builtin module_utils

Fixes https://github.com/ansible/ansible/issues/80524

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
module_utils

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

File: changelogs/fragments/remove-python3.5.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+removed_features:
+  - module_utils/basic.py - removed Python 3.5 as a supported remote version. Python 2.7 or Python 3.6+ is now required.
```
Comment (mattclay) [inline changelogs/fragments/remove-python3.5.yml]:
```suggestion
  - module_utils/basic.py - Removed Python 3.5 as a supported remote version. Python 2.7 or Python 3.6+ is now required.
```

File: lib/ansible/executor/task_executor.py [hunk 2]
```diff
@@ -1249,8 +1249,7 @@ def start_connection(play_context, options, task_uuid):
     else:
         try:
             result = json.loads(to_text(stderr, errors='surrogate_then_replace'))
-        except getattr(json.decoder, 'JSONDecodeError', ValueError):
-            # JSONDecodeError only available on Python 3.5+
+        except (json.decoder.JSONDecodeError, ValueError):
```
Comment (mattclay) [inline lib/ansible/executor/task_executor.py]: This should probably go in a separate PR, since the only relation it has with Python 3.5 is the comment. The fallback behavior hasn't been needed since ansible-core 2.11 when Python 2.7 was still unofficially supported as a controller.

That said, when it is included, there's a simpler form to use:

```suggestion
        except json.decoder.JSONDecodeError:
```

**Review by mattclay — CHANGES_REQUESTED** [review]
There are a few more Python 3.5 references in remote-only code that can be removed:

https://github.com/ansible/ansible/blob/5e550b608632251e2e919d1d85644a901f0233af/lib/ansible/modules/pip.py#L64

https://github.com/ansible/ansible/blob/5e550b608632251e2e919d1d85644a901f0233af/lib/ansible/module_utils/basic.py#L1724-L1725

File: lib/ansible/module_utils/basic.py [hunk 3]
```diff
@@ -1722,7 +1722,7 @@ def atomic_move(self, src, dest, unsafe_writes=False):
                     #     file = _os.path.join(dir, pre + name + suf)
                     # TypeError: can't concat bytes to str
                     error_msg = ('Failed creating tmp file for atomic move.  This usually happens when using Python3 less than Python3.5. '
-                                 'Please use Python2.x or Python3.5 or greater.')
+                                 'Please use Python2.x or Python3.6 or greater.')
```
Comment (mattclay) [inline lib/ansible/module_utils/basic.py]: Since we're already enforcing the use of Python 2.7 or 3.6+ at the top of the file, it looks like we can remove the entire `TypeError` handler, rather than just updating the error message. In fact, we should have been able to do that quite some time ago, since we've required Python 2.x or 3.5+ since at least 2.12.

**Review by mattclay — CHANGES_REQUESTED** [review]

File: lib/ansible/module_utils/basic.py (hunk 3 above)
Comment (mattclay) [reply lib/ansible/module_utils/basic.py]: If you want to make the exception handler change in a separate PR (along with the `JSONDecodeError` change), that's fine too. Just let me know which way you'd like to handle it.

Comment (s-hertel) [reply lib/ansible/module_utils/basic.py]: The "usually" in this error message, and the fact we've never supported Python < 3.5 made me hesitate to remove it. I think you're right, that it could have been removed back in 2.5 when we added the check at the top of the file. We didn't enforce the version for 2.3 which is when this was added. Want me to put the JSONDecodeError change back if I remove this section?

File: lib/ansible/executor/task_executor.py (hunk 2 above)
Comment (s-hertel) [reply lib/ansible/executor/task_executor.py]: Added back. Good spotting, this was a weird typo :)

**Review by mattclay — APPROVED** [review]
